Use bump ranges instead of free list in JSC allocator

ddegazio opened the pull request:
#### 2fbb3152165bdcb84f16b795863476c1816428a2
<pre>
Use bump ranges instead of free list in JSC allocator
<a href="https://bugs.webkit.org/show_bug.cgi?id=255798">https://bugs.webkit.org/show_bug.cgi?id=255798</a>
rdar://108377867

Reviewed by Yusuke Suzuki and Mark Lam.

Replaces the JSC free list allocator with a list of free intervals.
Each interval represents a contiguous block of dead cells, and within
each interval we can bump-allocate. This should mean we bump-allocate
more often, and potentially speed up sweeping by nature of installing
fewer free list cells, especially for use cases where we don&apos;t often
manage to totally empty blocks currently.

* Source/JavaScriptCore/heap/FreeList.cpp:
(JSC::FreeList::clear):
(JSC::FreeList::initialize):
(JSC::FreeList::contains const):
(JSC::FreeList::dump const):
(JSC::FreeList::initializeList): Deleted.
(JSC::FreeList::initializeBump): Deleted.
* Source/JavaScriptCore/heap/FreeList.h:
(JSC::FreeCell::scramble):
(JSC::FreeCell::descramble):
(JSC::FreeCell::makeLast):
(JSC::FreeCell::setNext):
(JSC::FreeCell::decode):
(JSC::FreeCell::offsetOfScrambledBits):
(JSC::FreeList::allocationWillFail const):
(JSC::FreeList::isSentinel):
(JSC::FreeList::offsetOfHead):
(JSC::FreeList::offsetOfPayloadStart):
(JSC::FreeList::offsetOfPayloadEnd):
(JSC::FreeList::head const):
(JSC::FreeCell::next const): Deleted.
(JSC::FreeCell::offsetOfScrambledNext): Deleted.
(JSC::FreeList::offsetOfScrambledHead): Deleted.
(JSC::FreeList::offsetOfRemaining): Deleted.
* Source/JavaScriptCore/heap/FreeListInlines.h:
(JSC::FreeList::allocate):
(JSC::FreeList::forEach const):
* Source/JavaScriptCore/heap/MarkedBlockInlines.h:
(JSC::MarkedBlock::Handle::specializedSweep):
* Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
(JSC::AssemblyHelpers::jitAssertTagsInPlace):
(JSC::AssemblyHelpers::emitExceptionCheck):
(JSC::AssemblyHelpers::emitNonPatchableExceptionCheck):
(JSC::AssemblyHelpers::loadProperty):
(JSC::AssemblyHelpers::storeProperty):
(JSC::AssemblyHelpers::emitAllocateWithNonNullAllocator):
(JSC::AssemblyHelpers::emitAllocateVariableSized):
(JSC::AssemblyHelpers::restoreCalleeSavesFromEntryFrameCalleeSavesBuffer):
(JSC::AssemblyHelpers::emitRestoreCalleeSavesFor):

Canonical link: <a href="https://commits.webkit.org/263313@main">https://commits.webkit.org/263313@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/80cb2596019cdb7278d81ac867efb8dde0cb2d25

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4224 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4345 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4463 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5694 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4470 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4452 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4310 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4688 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4285 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4462 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3814 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5688 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1957 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3791 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/6105 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/3523 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3784 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/3857 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/5379 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/4020 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4264 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/3461 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/4321 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3774 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3789 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1070 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1035 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/7866 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/4418 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4118 "Built successfully") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/3/builds/1178 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/heap/FreeList.h:
```diff
@@ -27,36 +27,47 @@
+    static ALWAYS_INLINE uint64_t scramble(int32_t offsetToNext, uint32_t lengthInBytes, uintptr_t secret)
     {
-        return bitwise_cast<uintptr_t>(cell) ^ secret;
+        ASSERT((uint64_t(lengthInBytes) << 32 | offsetToNext));
```
WebKit style would say to use a `static_cast<uint64_t>(...)` instead of `uint64_t(...)`.

ddegazio replied:
Resolved in latest push.

MenloDorian commented on Source/JavaScriptCore/heap/FreeList.h:
```diff
@@ -27,36 +27,47 @@
-        return bitwise_cast<FreeCell*>(cell ^ secret);
+        static_assert(WTF::isPowerOfTwo(sizeof(FreeCell))); // Make sure this division isn't super costly.
+        uint64_t descrambledBits = scrambledBits ^ secret;
+        return { int32_t(uint32_t(descrambledBits)), uint32_t(descrambledBits >> 32u) };
```
Ditto with casting style.

ddegazio replied:
Resolved in latest push.

MenloDorian commented on Source/JavaScriptCore/heap/FreeListInlines.h:
```diff
@@ -33,35 +33,45 @@ namespace JSC {
-        m_remaining = remaining;
-        return bitwise_cast<HeapCell*>(m_payloadEnd - remaining - cellSize);
+    unsigned cellSize = m_cellSize;
+    if (LIKELY(m_payloadStart != m_payloadEnd)) {
```
Let's make this `m_payloadStart < m_payloadEnd`.

ddegazio replied:
Resolved in latest push.

MenloDorian commented on Source/JavaScriptCore/heap/FreeListInlines.h:
```diff
@@ -33,35 +33,45 @@ namespace JSC {
-    FreeCell* result = head();
-    if (UNLIKELY(!result))
+    FreeCell* cell = head();
+    if (isSentinel(cell))
```
UNLIKELY?

ddegazio replied:
Added this back in latest push, this should be pretty unlikely.

ddegazio replied:
Added this back in latest push, this should be pretty unlikely.

MenloDorian commented on Source/JavaScriptCore/heap/FreeListInlines.h:
```diff
@@ -33,35 +33,45 @@ namespace JSC {
+            payloadStart = bitwise_cast<char*>(cell);
+            payloadEnd = payloadStart + lengthInBytes;
+            ASSERT(payloadEnd - payloadStart < (ptrdiff_t)(16 * KB));
+            cell = bitwise_cast<FreeCell*>(bitwise_cast<char*>(cell) + offsetToNext);
```
Just use `payloadStart` instead of `bitwise_cast<char*>(cell)` here.

ddegazio replied:
Resolved in latest push.

MenloDorian commented on Source/JavaScriptCore/heap/FreeList.cpp:
```diff
@@ -39,53 +39,50 @@ FreeList::~FreeList()
     m_secret = secret;
-    m_payloadEnd = nullptr;
-    m_remaining = 0;
+    m_payloadStart = reinterpret_cast<char*>(start);
```
Use `bitwise_cast` here instead of `reinterpret_cast`.  You can also initialize `m_head` in terms of `m_payloadStart` instead of having to do `bitwise_cast<char*>(start)` 2 times (not that it really matters for perf, but I think it's good to clearly point out the the head is an offset from the start of the current payload / interval.

ddegazio replied:
Resolved in latest push.

MenloDorian commented on Source/JavaScriptCore/heap/FreeList.h:
```diff
@@ -94,12 +105,12 @@ class FreeList {
-    uintptr_t m_scrambledHead { 0 };
-    uintptr_t m_secret { 0 };
+    char* m_payloadStart { nullptr };
     char* m_payloadEnd { nullptr };
```
Let's name these `m_intervalStart` and `m_intervalEnd` instead to match the algorithm.  The `payload` term is a bit ambiguous.

ddegazio replied:
Resolved in latest push.

MenloDorian commented on Source/JavaScriptCore/heap/FreeList.h:
```diff
@@ -94,12 +105,12 @@ class FreeList {
+    char* m_payloadStart { nullptr };
     char* m_payloadEnd { nullptr };
-    unsigned m_remaining { 0 };
+    FreeCell* m_head { reinterpret_cast_ptr<FreeCell*>(1) };
```
Let's name this `m_nextInterval` instead of `m_head`.  `m_head` implies this is where we start, which is ambiguous if one is thinking of where we start allocating.  `m_nextInterval` (paired with `m_intervalStart` and `m_intervalEnd`) makes it clear what this is really for.

ddegazio replied:
Resolved in latest push.

MenloDorian commented on Source/JavaScriptCore/heap/FreeList.cpp:
```diff
@@ -39,53 +39,50 @@ FreeList::~FreeList()
+    while (!isSentinel(candidate)) {
+        auto [offsetToNext, lengthInBytes] = candidate->decode(m_secret);
+        char* end = bitwise_cast<char*>(candidate + 1);
+        char* start = end - lengthInBytes;
```
This is wrong.  `start` should be `bitwise_cast<char*>(candidate)`, and `end` should be `start + lengthInBytes`.  You're doing this computation in so many places that maybe you should have an ALWAYS_INLINE helper function to do them.

ddegazio replied:
Fixed the `contains()` method in latest push, but I still need to refactor this code duplication out.

ddegazio replied:
Refactored to use a single `FreeCell::advance()` function that loads and decodes the current interval, sets up the interval start and end, and computes the next next interval in a single call. Should resolve the code duplication issue.

Constellation commented on Source/JavaScriptCore/heap/FreeList.cpp:
```diff
@@ -39,53 +39,49 @@ FreeList::~FreeList()
-    m_scrambledHead = 0;
+    m_intervalStart = nullptr;
+    m_intervalEnd = nullptr;
+    m_nextInterval = reinterpret_cast_ptr<FreeCell*>(1);
```
I like the thing like, `m_nextInterval = bitwise_cast<FreeCell*>(static_cast<uintptr_t>(1));`

ddegazio replied:
Changed in most recent patch.

Constellation commented on Source/JavaScriptCore/heap/FreeList.h:
```diff
@@ -27,36 +27,45 @@
-    
-    static FreeCell* descramble(uintptr_t cell, uintptr_t secret)
+
+    static ALWAYS_INLINE std::pair<int32_t, uint32_t> descramble(uint64_t scrambledBits, uint64_t secret)
```
Probably using `std::tuple` would be better for new code (destructuring assignment works with that too).

ddegazio replied:
Changed in latest push.

Constellation commented on Source/JavaScriptCore/heap/FreeList.h:
```diff
@@ -27,36 +27,45 @@
-    
-    static ptrdiff_t offsetOfScrambledNext() { return OBJECT_OFFSETOF(FreeCell, scrambledNext); }
+
+    ALWAYS_INLINE std::pair<int32_t, uint32_t> decode(uint64_t secret)
```
Ditto.

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -607,44 +607,113 @@ void AssemblyHelpers::emitAllocateWithNonNullAllocator(GPRReg resultGPR, const J
+    addUnsignedRightShift64(resultGPR, scratchGPR, TrustedImm32(32), scratchGPR);
+    storePtr(scratchGPR, Address(allocatorGPR, LocalAllocator::offsetOfFreeList() + FreeList::offsetOfIntervalEnd()));
+    jump(bumpLabel);
+#elif CPU(X86_64) || CPU(X86)
```
`CPU(X86)` is not necessary since we are not supporting 32bit x86 JIT.

ddegazio replied:
Removed in latest push.

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -607,44 +607,113 @@ void AssemblyHelpers::emitAllocateWithNonNullAllocator(GPRReg resultGPR, const J
     // - We *can* use RegisterSetBuilder::macroScratchRegisters on ARM.
 
-    Jump popPath;
+    Jump popPath, zeroPath;
```
Let's have `Jump zeroPath`. We normally do not use comma separated declarations.

ddegazio replied:
Resolved in latest push.

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -607,44 +607,113 @@ void AssemblyHelpers::emitAllocateWithNonNullAllocator(GPRReg resultGPR, const J
+    zeroPath = branchTestPtr(ResultCondition::NonZero, resultGPR, TrustedImm32(1));
+    load32(Address(allocatorGPR, LocalAllocator::offsetOfFreeList() + FreeList::offsetOfSecret()), scratchGPR);
+    xor32(Address(resultGPR, FreeCell::offsetOfScrambledBits()), scratchGPR); // Lower 32 bits -> offset to next interval
+    if constexpr (is64Bit())
```
Since we do not support 32bit x86, you can always assume `is64Bit()` is true here.

ddegazio replied:
Resolved in latest push.

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -607,44 +607,113 @@ void AssemblyHelpers::emitAllocateWithNonNullAllocator(GPRReg resultGPR, const J
+        if constexpr (is64Bit())
+            add64(TrustedImm32(allocator.allocator().cellSize()), Address(allocatorGPR, LocalAllocator::offsetOfFreeList() + FreeList::offsetOfIntervalStart()));
+        else
+            add32(TrustedImm32(allocator.allocator().cellSize()), Address(allocatorGPR, LocalAllocator::offsetOfFreeList() + FreeList::offsetOfIntervalStart()));
```
I think `FreeList::offsetOfIntervalStart()` is not updated yet if we come from popPath, correct? So, it is incrementing old value.

ddegazio replied:
Good catch! Should be resolved in the most recent push.

MenloDorian commented on Source/JavaScriptCore/heap/FreeList.h:
```diff
@@ -27,36 +27,53 @@
 
 struct FreeCell {
-    static uintptr_t scramble(FreeCell* cell, uintptr_t secret)
+    static ALWAYS_INLINE uint64_t scramble(int32_t offsetToNext, uint32_t lengthInBytes, uintptr_t secret)
```
`secret` should be `uint64_t`.  `uintptr_t` would be `uint32_t` on  32-bit platforms.

MenloDorian commented on Source/JavaScriptCore/heap/FreeList.h:
```diff
@@ -66,10 +83,9 @@ class FreeList {
+    JS_EXPORT_PRIVATE void initialize(FreeCell* head, uint64_t secret, unsigned bytes);
     
-    bool allocationWillFail() const { return !head() && !m_remaining; }
+    bool allocationWillFail() const { return isSentinel(nextInterval()) && m_intervalStart >= m_intervalEnd; }
```
I presume we'll tend to have an interval to allocate out of.  So, let's favor this by putting `m_intervalStart >= m_intervalEnd` before `isSentinel(nextInterval())`.

MenloDorian commented on Source/JavaScriptCore/heap/FreeList.h:
```diff
@@ -94,12 +111,12 @@ class FreeList {
+    char* m_intervalStart { nullptr };
+    char* m_intervalEnd { nullptr };
+    FreeCell* m_nextInterval { bitwise_cast<FreeCell*>(static_cast<uintptr_t>(1)) };
     uintptr_t m_secret { 0 };
```
`m_secret` should be a `uint64_t`.

MenloDorian commented on Source/JavaScriptCore/heap/FreeListInlines.h:
```diff
@@ -33,36 +33,49 @@ namespace JSC {
+    FreeCell::advance(m_secret, m_nextInterval, m_intervalStart, m_intervalEnd);
     
-    m_scrambledHead = result->scrambledNext;
+    // We assume we don't create any empty intervals; so there should be enough space remaining
```
Let's remove the "assume we".  It's a fact that we don't create empty intervals.

Constellation commented on Source/JavaScriptCore/heap/FreeListInlines.h:
```diff
@@ -33,36 +33,49 @@ namespace JSC {
+            break;
+
+        FreeCell::advance(m_secret, cell, intervalStart, intervalEnd);
     }
```
Maybe, this is cleaner?
```
while (true) {
    for (; intervalStart < intervalEnd; intervalStart += m_cellSize)
        func(bitwise_cast<HeapCell*>(intervalStart));
            
    // If we explore the whole interval and the cell is the sentinel value, though, we should
    // immediately exit so we don't decode anything out of bounds.
    if (isSentinel(cell))
        break;

    FreeCell::advance(m_secret, cell, intervalStart, intervalEnd);
}
```

ddegazio replied:
Oh, yes I agree, I'll quickly make that change.

Constellation reviewed (approved):
r=me

MenloDorian commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -608,43 +608,110 @@ void AssemblyHelpers::emitAllocateWithNonNullAllocator(GPRReg resultGPR, const J
+    popPath.link(this);
+    loadPairPtr(allocatorGPR, TrustedImm32(LocalAllocator::offsetOfFreeList() + FreeList::offsetOfNextInterval()), resultGPR, scratchGPR);
+    zeroPath = branchTestPtr(ResultCondition::NonZero, resultGPR, TrustedImm32(1));
+    xorPtr(Address(resultGPR, FreeCell::offsetOfScrambledBits()), scratchGPR);
```
Let's use `xor64` since `scrambledBits` is a `uint64_t`.

MenloDorian commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -608,43 +608,110 @@ void AssemblyHelpers::emitAllocateWithNonNullAllocator(GPRReg resultGPR, const J
+    // and extracting interval information use less instructions.
+
+    // Bump allocation (fast path)
+    loadPairPtr(allocatorGPR, TrustedImm32(LocalAllocator::offsetOfFreeList() + FreeList::offsetOfIntervalStart()), resultGPR, scratchGPR);
```
You should add a `static_assert` here that `FreeList::offsetOfIntervalEnd() == FreeList::offsetOfIntervalStart() + sizeof(char*)` because this `loadPairPtr` relies on it.

ddegazio replied:
I guess we can't `static_assert` this, but a normal `RELEASE_ASSERT` here shouldn't be a problem. I'll add one.

MenloDorian commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -608,43 +608,110 @@ void AssemblyHelpers::emitAllocateWithNonNullAllocator(GPRReg resultGPR, const J
 
+    // Get next interval (slower path)
+    popPath.link(this);
+    loadPairPtr(allocatorGPR, TrustedImm32(LocalAllocator::offsetOfFreeList() + FreeList::offsetOfNextInterval()), resultGPR, scratchGPR);
```
You should `static_assert` here that `FreeList:: offsetOfSecret() = FreeList::offsetOfNextInterval() + sizeof(FreeCell*)` because this `loadPairPtr` relies on it.

ddegazio replied:
Same as above.

MenloDorian commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -608,43 +608,110 @@ void AssemblyHelpers::emitAllocateWithNonNullAllocator(GPRReg resultGPR, const J
+
+    // Bump allocation (fast path)
+    loadPairPtr(allocatorGPR, TrustedImm32(LocalAllocator::offsetOfFreeList() + FreeList::offsetOfIntervalStart()), resultGPR, scratchGPR);
+    popPath = branchPtr(RelationalCondition::Equal, resultGPR, scratchGPR);
```
nit: Let's use `RelationalCondition::AboveOrEqual` instead.

MenloDorian reviewed (approved):
r=me.

I think with this change, `Options::useBumpAllocator()` no longer has meaning.  But you can remove that later in a follow up patch.

ddegazio committed:
Use bump ranges instead of free list in JSC allocator
https://bugs.webkit.org/show_bug.cgi?id=255798
rdar://108377867

Reviewed by Yusuke Suzuki and Mark Lam.

Replaces the JSC free list allocator with a list of free intervals.
Each interval represents a contiguous block of dead cells, and within
each interval we can bump-allocate. This should mean we bump-allocate
more often, and potentially speed up sweeping by nature of installing
fewer free list cells, especially for use cases where we don't often
manage to totally empty blocks currently.

* Source/JavaScriptCore/heap/FreeList.cpp:
(JSC::FreeList::clear):
(JSC::FreeList::initialize):
(JSC::FreeList::contains const):
(JSC::FreeList::dump const):
(JSC::FreeList::initializeList): Deleted.
(JSC::FreeList::initializeBump): Deleted.
* Source/JavaScriptCore/heap/FreeList.h:
(JSC::FreeCell::scramble):
(JSC::FreeCell::descramble):
(JSC::FreeCell::makeLast):
(JSC::FreeCell::setNext):
(JSC::FreeCell::decode):
(JSC::FreeCell::offsetOfScrambledBits):
(JSC::FreeList::allocationWillFail const):
(JSC::FreeList::isSentinel):
(JSC::FreeList::offsetOfHead):
(JSC::FreeList::offsetOfPayloadStart):
(JSC::FreeList::offsetOfPayloadEnd):
(JSC::FreeList::head const):
(JSC::FreeCell::next const): Deleted.
(JSC::FreeCell::offsetOfScrambledNext): Deleted.
(JSC::FreeList::offsetOfScrambledHead): Deleted.
(JSC::FreeList::offsetOfRemaining): Deleted.
* Source/JavaScriptCore/heap/FreeListInlines.h:
(JSC::FreeList::allocate):
(JSC::FreeList::forEach const):
* Source/JavaScriptCore/heap/MarkedBlockInlines.h:
(JSC::MarkedBlock::Handle::specializedSweep):
* Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
(JSC::AssemblyHelpers::jitAssertTagsInPlace):
(JSC::AssemblyHelpers::emitExceptionCheck):
(JSC::AssemblyHelpers::emitNonPatchableExceptionCheck):
(JSC::AssemblyHelpers::loadProperty):
(JSC::AssemblyHelpers::storeProperty):
(JSC::AssemblyHelpers::emitAllocateWithNonNullAllocator):
(JSC::AssemblyHelpers::emitAllocateVariableSized):
(JSC::AssemblyHelpers::restoreCalleeSavesFromEntryFrameCalleeSavesBuffer):
(JSC::AssemblyHelpers::emitRestoreCalleeSavesFor):

Canonical link: https://commits.webkit.org/263313@main
